feat(md-chips): added validation for ng-required

IMM0rtalis opened the pull request:
## PR Checklist
Please check that your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/angular/material/blob/master/.github/CONTRIBUTING.md#-commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Infrastructure changes
[x] Other... Please describe: In my opinion it is more a bug than a feature because this should be a default behavior
```

## What is the current behavior?
ng-required not working with md-chips

Issue Number: #11124


## What is the new behavior?
ng-required works with md-chips

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No (I would say no)
```

## Other information
Pull-Request is mainly a follow-up from #8034. I had the same idea in how to add this but found this not followed/merged pull-request by @Emeegeemee .
So nearly everything of the code got reviewed once by @DevVersion & @crisbeto . I just changed the naming a bit.

Splaktar commented on src/components/chips/demoValidation/index.html:
```diff
@@ -0,0 +1,27 @@
+      </div>
+    </form>
+  </md-content>
+</div>
```
Blank line at end of file please.

Splaktar commented on src/components/chips/demoValidation/script.js:
```diff
@@ -0,0 +1,11 @@
+    this.selectedFruit = [];
+    this.selectedVegetables = [];
+  }
+})();
```
Blank line at end of file please.

IMM0rtalis replied:
Is this fixed @Splaktar ? Because in the code it seems like, but just the one comment of you got hidden, this here not.

Splaktar replied:
Yep, looks good now.

Splaktar commented on src/components/chips/demoValidation/index.html:
```diff
@@ -0,0 +1,27 @@
+      </md-chips>
+      <div ng-messages="fruitForm.fruit.$error">
+        <div ng-message="required">At least one fruit is required</div>
+      </div>
```
These don't show up red because the `chips-theme.scss` is lacking styles for `md-invalid`. Can you please review the `input-theme.scss` and implement something similar for chips as part of this PR?

Splaktar replied:
`input.scss` also uses a `md-input-message-animation` class to animate the message in and to set a smaller line height and font size (among other things). Can we please add those styles to `chips.scss` so that they will be similar when used in a form together?

Splaktar replied:
The error message is showing up even when the input is pristine and not dirty. For input, this isn't the case. Can you please look into this?

IMM0rtalis replied:
Ok, I found out what md-input-container is doing: It adds a special animation-class `md-input-messages-animation` (https://github.com/angular/material/blob/v1.1.7/src/components/input/input.js#L840) and does some stuff with it. I see two ways here. A: make the demos with md-chips inside md-input-container and assure that this have the same look&feel. B: Add a chips-special-class and add styles there. 
But B is not really following the ways it is made atm. ng-messages is just styled inside certain containers. So just A is left, but are chips belonging to md-input-container? - Is there any rule which forbids it?

Splaktar replied:
Yeah, I can see how we wouldn't want to style `ng-messages` that aren't contained in an AngularJS Material component. I've asked @DevVersion for some input and created a [CodePen](https://codepen.io/Splaktar/pen/QQmwop) to demonstrate how `ng-messages` are styled with chips today (with `md-max-chips`) for reference since there is no example in the demos.

IMM0rtalis commented:
@Splaktar you labeled this PR and the related issue with "bug". Is it a bug or feature in your opinion? - because then I should adjust the commit-message-prefix.

Splaktar commented:
Yeah, I think you are right. I'll updated this to `enhancement`. Using `feat` in the commit message is good.

DevVersion commented:
I think just implementing the logic for supporting the `required` validator is fine here. That's the main intention of the PR and the design of error messages for chips is more a separate issue in my opinion.

For now, we could just have some custom CSS in the demo and maybe show a small caption under the chips demo that explains that there is no default CSS for error messages yet.

Placing chips inside of the  `<md-input-container>` element makes sense, because the chips can be also considered as an input. Since making the input container more abstract will likely result in breaking changes, I'd vote for another container like `<md-form-field>`.

The `<md-form-field>` and `<md-input-container>` would then share somehow the CSS for the error messages. e.g. using a class like `.md-error-messages-container`. The actual directives for animations could then be also updated to work in either `form-field` or `input-container`.

As you see, this turns out to be a whole new issue, so I'd say that this should be handled in a follow-up.

Splaktar commented:
@DevVersion thank you very much. That makes a lot of sense.

@IMM0rtalis can you please look at https://github.com/angular/material/pull/11125#discussion_r168939210 and then add the styles to the demo along with a note about developers needing to style the `ng-messages` for chips? This note could be similar to the note in the Static Chips Demo.

IMM0rtalis commented:
@Splaktar I added the note and example-styles and conditional ng-show of error-messages in the demo.
While doing this I ran into problems with the dirty & touched-state not bubbled up from the input so I added a watcher. 
In general I figured out that no watcher was deregistered in the chipsController so I also added the destruction of some $watchers for onDestroy.

salimane commented:
@IMM0rtalis is there a way to use this branch in a project ? 

I tried without success : 

```json
"devDependencies": {
    "angular-material-source": "git+https://github.com/IMM0rtalis/material.git#c60f055728e1102afa7bb5e9d6e18df021000941"
}
```

IMM0rtalis commented:
@salimane not sure, but until the next version with this included is published you could try a hacky way (we had to do it for now)

```javascript
/**
 * Fix MdChipsController
 *  - fixes problem with not validating required-field (fix in v1.1.8)
 *
 * @see https://github.com/angular/material/pull/11125
 */
(function() {
    function updateMdChipsController(MdChipsCtrl) {

        /**
         * Updates the validity properties for the ngModel.
         */
        MdChipsCtrl.prototype.validateModel = function() {
            this.ngModelCtrl.$setValidity('md-max-chips', !this.hasMaxChipsReached());
            this.ngModelCtrl.$validate(); // rerun any registered validators
        };

        /**
         * Configures the required interactions with the ngModel Controller.
         * Specifically, set {@code this.items} to the {@code NgModelCtrl#$viewVale}.
         * @param ngModelCtrl
         */
        MdChipsCtrl.prototype.configureNgModel = function(ngModelCtrl) {
            this.ngModelCtrl = ngModelCtrl;

            var self = this;

            // in chips the meaning of $isEmpty changes
            ngModelCtrl.$isEmpty = function(value) {
                return !value || value.length === 0;
            };

            ngModelCtrl.$render = function() {
                // model is updated. do something.
                self.items = self.ngModelCtrl.$viewValue;
            };
        };
    }

    var invokes = angular.module('material.components.chips')._invokeQueue,
        i;

    for(i=0; i < invokes.length; ++i) {
        if(invokes[i][0] === "$controllerProvider" && invokes[i][2][0] === "MdChipsCtrl") {
            updateMdChipsController(invokes[i][2][1]);
        }
    }

})();
```

salimane commented:
thanks @IMM0rtalis .  what is the connection between the current repo and this one https://github.com/angular/bower-material .
do we need the same changes over there ?

Thanks

Splaktar commented:
@salimane bower-material is the repo where the releases live. This is what gets installed by bower or NPM. Everytime we merge something into master (including making a release), a new, installable build is generated in bower-material and available to Bower and NPM.

Splaktar commented on src/components/chips/js/chipsController.js:
```diff
@@ -170,18 +177,40 @@ MdChipsCtrl.prototype.init = function() {
 
   // Set the wrapper ID
-  ctrl.wrapperId = '_md-chips-wrapper-' + ctrl.$mdUtil.nextUid();
+  this.wrapperId = '_md-chips-wrapper-' + this.$mdUtil.nextUid();
```
What's the benefit of the changes on this line?

IMM0rtalis replied:
Actually I do not see a point in using a helper-var instead of `this` if `this` is usable in a place. In my opinion it reduces the readability of code. And if the helper-var for encapsulated inner functions is called `ctrl`, `self` or `that` is in my opinion too much a matter of taste (otherwise I would have replaced `ctrl` with more widely used helpers in this repository)

Splaktar replied:
Ah OK. 👍

salimane commented:
@Splaktar thanks for reply. 

Do you think there is way to have a branch in https://github.com/angular/bower-material that will use the changes in the current pull-request ? 

Thanks

Splaktar commented on src/components/chips/js/chipsController.js:
```diff
@@ -170,18 +177,40 @@ MdChipsCtrl.prototype.init = function() {
+  );
+
+  this.deRegister.push(
+    this.$scope.$watch('$mdChipsCtrl.items', function() {
```
The comment here is the same as above. What's the reason that this new `$watch` is needed?

IMM0rtalis replied:
You are just pointing out the replacement of `ctrl` with `this`, right? Then see comment above^^

Splaktar replied:
Nope, I'm asking about the two comments "Make sure our input and wrapper have the correct ARIA attributes" and the existing `$watchCollection` versus the new `$watch`.

IMM0rtalis replied:
Just added the unwatcher-handling and replaced `ctrl` with `this`
![screenshot_20180226_101737](https://user-images.githubusercontent.com/1506919/36662386-8b9466e6-1ade-11e8-96b3-bf2d89723b56.png)

IMM0rtalis replied:
Wait... this is duplicated. Fix it asap

Splaktar commented:
@salimane not until we get this merged.

Splaktar commented:
I asked @DevVersion to take one last look at this. Then I think that we can send it to presubmit.

DevVersion commented on src/components/chips/js/chipsController.js:
```diff
@@ -170,18 +177,32 @@ MdChipsCtrl.prototype.init = function() {
+  var $destroyFn;
+  while (($destroyFn = this.deRegister.pop())) {
+    $destroyFn.call(this);
+  }
```
Why is this made so complicated? What about just `this.unregisterFunctions.forEach(fn => fn());`?

IMM0rtalis replied:
Because it is ES6, right? Is this repo require es6 ?

Splaktar replied:
ES6 is fine in the build scripts and docs site, but not in the production library code (i.e. anything in `src/`).

DevVersion replied:
Yeah, using the `ES5` `forEach` method should be still a benefit.

DevVersion commented on src/components/chips/js/chipsController.js:
```diff
@@ -716,6 +744,43 @@ MdChipsCtrl.prototype.onInputBlur = function () {
+    this.deRegister.push(
+      this.$scope.$watch(
+        function() {
+          return ngModelCtrl.$touched;
```
You can combine this single watch into two. e.g.

```js
this.$scope.$watch(() => {
  return { touched: inputNgModelCtrl.$touched, dirty: inputNgModelCtrl.$dirty };
}, XXXX);
```

IMM0rtalis replied:
But dont I need $watchCollection for this? And then separate afterwards anyway?

DevVersion replied:
Don't think a `$watchCollection` will work here, because it requires us to watch properties on the `$scope`. Using `$watch` and just having a different watch condition should work.

https://stackoverflow.com/questions/17550947/how-to-watch-multiple-variables

IMM0rtalis replied:
If I do as suggested I run in an endless digest-cycle with karma. What speaks against having to simple watcher?

Splaktar replied:
Let's leave it as is, thanks for giving it a try.

IMM0rtalis commented:
@tinayuangao or @jelbourn or @andrewseguin or @josephperrott or @mmalerba 
When I understoof correct you are the ones who can/should merge it, right? - Since here is no activity for like 8-10 days I thought it maybe got lost between all the issues. Can sb pls merge it?

Splaktar commented:
@IMM0rtalis there is no need to mention all of the caretakers. The issue can't be merged until it passes internal presubmit testing.

@tinayuangao started the presubmits on this on 2/27, but I haven't received the results from that yet. I've pinged her for that information.

Thank you for the reminder.

Splaktar commented:
Presubmits failed. The failures are being investigated. I'll provide more details when they are available.

IMM0rtalis commented:
Ahh ok. I thought they are the ones who should merge it (https://github.com/angular/material/blob/master/docs/guides/COMMIT_LEVELS.md) and since there was no update...  But yeah, they are often not passed I have the feeling :D

Splaktar commented:
Yep, they will merge it when presubmits get sorted out and any internal failures are fixed (by them). They rotate shifts each week, so mentioning them all isn't ideal. I know who is on shift each week and I work with them to get the merges run through presubmit and merged. This process starts after I apply the `pr: merge ready` tag.

We recently updated many of the tag descriptions (new GitHub feature) to help make this process a bit more clear. I'll try to document a bit more of the process in the commit levels guide when I have some time.

jelbourn commented:
Passes presubmit

Splaktar commented:
This appears to have caused a regression related to adding chips onBlur. We're planning to revert it in https://github.com/angular/material/pull/11215 and do more investigation/testing.
